src/libopensc/reader-pcsc.c

```diff
@@ -376,24 +376,44 @@ static int refresh_attributes(sc_reader_t *reader)
 
 	if (rv != SCARD_S_SUCCESS) {
 		if (rv == (LONG)SCARD_E_TIMEOUT) {
-			/* Timeout, no change from previous recorded state. Make sure that
-			 * changed flag is not set. */
-			reader->flags &= ~SC_READER_CARD_CHANGED;
+			DWORD readers_len = 0, cstate = 0, prot, atr_len = SC_MAX_ATR_SIZE;
+			unsigned char atr[SC_MAX_ATR_SIZE];
+
 			/* Make sure to preserve the CARD_PRESENT flag if the reader was
 			 * reattached and we called the refresh_attributes too recently */
 			if (priv->reader_state.dwEventState & SCARD_STATE_PRESENT) {
 				reader->flags |= SC_READER_CARD_PRESENT;
 			}
+
+			rv = priv->gpriv->SCardStatus(priv->pcsc_card, NULL, &readers_len, &cstate, &prot, atr, &atr_len);
+			if (priv->pcsc_card != 0
```

Comment:
Maybe you can test if `priv->pcsc_card != 0` **before** the call to `SCardStatus()` to note waste CPU cycles?

A `hCard` set to 0 is invalid for pcsc-lite https://github.com/LudovicRousseau/PCSC/blob/master/src/winscard.c#L1254-L1256. But I don't know for the other WinSCard implementations (Windows and macOS).